lib/daru/vector.rb

```diff
@@ -760,6 +760,43 @@ def replace_nils! replacement
       self
     end
 
+    # Rolling fillna
+    # replace all Float::NAN and NIL values with the preceeding or following value
+    #
+    # @param [Symbol] (:forward, :backward) whether replacement value is preceeding or following
+    #
+    # @example
+    #
+    # dv = Daru::Vector.new([1, 2, 1, 4, nil, Float::NAN, 3, nil, Float::NAN])
+    #
+    # 2.3.3 :068 > dv.rolling_fillna(:forward)
+    # => #<Daru::Vector(9)>
+    # 0   1
+    # 1   2
+    # 2   1
+    # 3   4
+    # 4   4
+    # 5   4
+    # 6   3
+    # 7   3
+    # 8   3
+    #
+    def rolling_fillna!(direction=:forward)
+      missing_idxs = indexes(*Daru::MISSING_VALUES)
+      valid_idxs = Array(0...dv.size) - missing_idxs
+      valid_idxs, comp = direction == :forward ? [valid_idxs.reverse, :<] : [valid_idxs, :>]
+      missing_idxs.each do |idx|
+        repl_idx = valid_idxs.find { |i| i.public_send(comp, idx) }
+        repl = repl_idx ? self[repl_idx] : 0
+        self[idx] = repl
+      end
```

Comment:
Don't you think it could be done simpler? I am too lazy to write working code, but idea is (pseudocode):
```ruby
enumerator = direction == :forward ? each : reverse_each
prev_valid = nil
enumerator.each do |idx, val| 
  if valid?(val)
    prev_valid = val
  else
    self[idx] = prev_valid
  end
end
```

Reply:
Hmm I might be missing the idea but I think my method basically follows this pattern?

Reply:
I've meant this (works against your specs):
```ruby
    def rolling_fillna!(direction=:forward)
      enumerator = direction == :forward ? index : index.reverse_each
      last_valid_value = 0
      enumerator.each do |idx|
        if valid_value?(self[idx])
          last_valid_value = self[idx]
        else
          self[idx] = last_valid_value
        end
      end
    end

    def valid_value?(v)
      v.respond_to?(:nan?) && v.nan? || v.nil? ? false : true
    end
```
What's the difference? Let's look at initial version:
1. List of missing indexes is created: enumeration through the vector (performance) + array of indexes (memory)
2. List of valid indexes is created: array of them (memory, and probably performance substracting say 10k items arrays)
3. Missing indexes is enumerated, and for each of them we search through valid indexes array (performance)

Proposed version:
1. Only one enumeration of entire vector
2. No additional huge arrays
;)

And, last but not least, try this test:
```ruby
    context 'with non-default index' do
      subject do
        Daru::Vector.new(
          [Float::NAN, 2, 1, 4, nil, Float::NAN, 3, nil, Float::NAN],
          index: %w[a b c d e f g h i]
        )
      end

      before { subject.rolling_fillna!(direction: :backward) }

      it { is_expected.to eq Daru::Vector.new([2, 2, 1, 4, 3, 3, 3, 0, 0], index: %w[a b c d e f g h i]) }
    end
```
...and prepare to be surprised.

Reply:
Thanks for the detailed explanation - much clearer now, have changed.